### PR TITLE
change providers to requests

### LINKED ISSUE #155: Rename "providers" to "presentationRequests" or just  "requests"?
As came up in #71, "providers" might not be the most clear member name to use:

@timcappalli [proposed](https://github.com/WICG/digital-credentials/pull/71#issuecomment-2260675780):

```WebIDL
dictionary DigitalCredentialRequestOptions {
  sequence<DigitalCredentialsPresentationRequest> presentationRequests;
};
```

I'd be ok with just "requests" tho, but `presentationRequests` is also pretty clear.

### PR DESCRIPTION
**Pull request by marcoscaceres**
Closes #155 

The following tasks have been completed:

- [x] [Modified Web platform tests](https://github.com/WICG/digital-credentials/pull/165)

Implementation commitment:

- [x] WebKit - [issue](https://bugs.webkit.org/show_bug.cgi?id=279933), [issue](https://bugs.webkit.org/show_bug.cgi?id=279931) 
- [ ] Chromium (link to issue)
- [ ] Gecko (link to issue)

Documentation and checks

- [ ] Affects privacy
- [ ] Affects security
- [ ] Pinged MDN
- [ ] Updated Explainer

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/digital-credentials/pull/165.html" title="Last updated on Sep 18, 2024, 11:23 PM UTC (48f6041)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/digital-credentials/165/5c07f1f...48f6041.html" title="Last updated on Sep 18, 2024, 11:23 PM UTC (48f6041)">Diff</a>

### CODE REVIEW

**Comment by samuelgoto** [comment]
Something that occurred to me that feels a bit off is that we'd have a `requests` and `request` in the API. Here:

```javascript
await navigator.credentials.get({
  digital: {
     requests: [{      <-----------
       protocol: "openid4vp",
       request: {.      <----------- is this weird?
         nonce: "foo",
         client_id_metadata: "bar",
         // ... more stuff
       },
     }]
  }
});
```

**Comment by marcoscaceres** [comment]
Yeah, I spotted that too... I can live with it though. 

How about `.query:`?

**Comment by marcoscaceres** [comment]
Some alternatives:
* `presentationRequests` and `presentationDetails`
* `presentmentRequests` and `requestDetails`

**Comment by bc-pi** [comment]
To add to the a bit odd but not terrible observation, due to the layering and reuse of prior RFCs, the top level thing[1] of a `request` in a signed openid4vp request is also a `request`. See not permanent content at https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#appendix-A.3.2-3 but that would mean something like the below with `requests: ... request: ... request: "..."` which is pretty special. 

```
await navigator.identity.get({
  digital: {
    requests: [{
      protocol: "openid4vp",
      request: {
        request: "eyJhbGciOiJF..."
     }
    }]
  }
});
```

[1] I can't even keep track of what things at what layer are called anymore

**Comment by bc-pi** [comment]
Can someone remind me what the rational for the `requests: [...]` array under `digital` rather than a single request?

**Comment by bc-pi** [comment]
> Can someone remind me what the rational for the `requests: [...]` array under `digital` rather than a single request?

The layers and potential options for various cardinalities is making my little mind hurt. Kinda related is https://github.com/openid/OpenID4VP/issues/248 not to mention that whatever query language(s) in openid4vp can themselves express asking for multiple things and/or combinatorial options.

**Comment by samuelgoto** [comment]
> Can someone remind me what the rational for the requests: [...] array under digital rather than a single request?

Yeah, @bc-pi , that's one of the things that I think we may be worth revisiting: I'm not quite sure anymore we need the level of indirection of `providers`.

To answer your question, originally, `providers` was introduced because we thought that there would be more than 1 `protocol` used (e.g. the **same** request using OpenID4VP, VC API, something coming from ISO, etc) and that wallets would accept "one or the other". A similar level of indirection that we introduced for formats (e.g. mDocs and SD-JWTs).  

At this point, I think it is not clear to me that we would run into this situation: when a verifier makes a request, it doesn't have to connect with a variety of wallets that support different protocols.

**Comment by marcoscaceres** [comment]
I wouldn't mind dropping `.requests` entirely to be honest. CoreIDV on Apple's platform doesn't support requesting multiple, so we would just pick the first known type. 

We should then add a. static `.supportsRequestType()` to the `DigitalCredential` interface, or some such. Otherwise developers won't know if it will immediately fail.

**Comment by msporny** [comment]
@samuelgoto wrote:
> To answer your question, originally, providers was introduced because we thought that there would be more than 1 protocol used (e.g. the same request using OpenID4VP, VC API, something coming from ISO, etc) and that wallets would accept "one or the other". A similar level of indirection that we introduced for formats (e.g. mDocs and SD-JWTs).

Hmm, no, we definitely still have that requirement. For example, in CA DMV, we might have a future where mDoc is requested over a different protocol (or protocol version) than other VCs. We cannot presume that there will only be one protocol.

**Comment by marcoscaceres** [comment]
Right, @msporny, but the future format would need to be known to the browser, so "mdoc" would need to be a registered protocol (i.e., you can't sent up anything as that would expose what is installed on the device, so for user privacy you can't send "wallet-custom-format"). The developer can check and do the filtering on their most preferred **standardized** type there.

**Comment by marcoscaceres** [comment]
Moved the proposal here: https://github.com/WICG/digital-credentials/issues/168

**Comment by msporny** [comment]
> The developer can check and do the filtering on their most preferred **standardized** type there.

Yes, agreed. I wasn't making an argument against having a set of standardized protocols or standardized formats.

It sounded like @samuelgoto was saying: "We don't need to contemplate a set of standardized protocols that a verifier might use to request a credential, because there will only ever be one that a verifier will use to request a credential." ... and that's not true today and is unlikely to be true in the future (as much as we'd all like that to be true).

**Comment by marcoscaceres** [comment]
Ok, talked with @samuelgoto and he also convinced me that we should keep the array. 

for the member names, how about we do `{ requests: [{protocol: "openid4vp", data: {} }] }`.

**Comment by bc-pi** [comment]
> > Can someone remind me what the rational for the requests: [...] array under digital rather than a single request?
> 
> Yeah, @bc-pi , that's one of the things that I think we may be worth revisiting: I'm not quite sure anymore we need the level of indirection of `providers`.
> 
> To answer your question, originally, `providers` was introduced because we thought that there would be more than 1 `protocol` used (e.g. the **same** request using OpenID4VP, VC API, something coming from ISO, etc) and that wallets would accept "one or the other". A similar level of indirection that we introduced for formats (e.g. mDocs and SD-JWTs).
> 
> At this point, I think it is not clear to me that we would run into this situation: when a verifier makes a request, it doesn't have to connect with a variety of wallets that support different protocols.

Thanks for backstory @samuelgoto. My general intuition is that there are too many levels of indirection in all the various layers of all this. But that's easy to say and not so easy to clearly say what should be removed.

**Comment by bc-pi** [comment]
> Ok, talked with @samuelgoto and he also convinced me that we should keep the array.
> 
> for the member names, how about we do `{ requests: [ { protocol: "openid4vp", data: {} }] }`.
> 
> So `requests` and `data`.

I like `requests` and `data`, thanks @marcoscaceres!

**Review by samuelgoto — APPROVED** [review]
LGTM

This is a backwards incompatible change, so we are going to need to update the OpenID4VP spec as well as verifiers. It is still possible, but we shouldn't take this lightly.

Can we run this by the community group to make sure that we are on all board with this trade-off?

**Review by bc-pi — APPROVED** [review]

File: index.html
Comment (samuelgoto) [inline index.html:237]:
```suggestion
      the [=digital credential/data=] to be handled by the holder's software,
```

File: index.html
Comment (samuelgoto) [inline index.html:117]:
```suggestion
        [=digital credential/request=] by a [=verifier=].
```

File: index.html
Comment (samuelgoto) [inline index.html:107]:
```suggestion
        A presentation request is a request for a [=digital credential=] composed
```

File: index.html
Comment (marcoscaceres) [inline index.html:104]:
```suggestion
        <dfn data-dfn-for="digital credential">Presentation request</dfn>
```

File: index.html
Comment (marcoscaceres) [inline index.html:217]:
```suggestion
      The {{DigitalCredentialsRequest}} dictionary represents a [=digital credential/presentation
      request=]. It is used to specify an [=digital credential/exchange
```

Comment (samuelgoto) [inline index.html]:
```suggestion
      request=]. It is used to specify an [=digital credential/exchange
```

**Review by samuelgoto — APPROVED** [review]
LGTM

This is a backwards incompatible change, but a good one in case we still have time. Lets run this by the CG before merging.

**Comment by samuelgoto** [comment]
Just for transparency, the other alternatives that we considered for `data` was `payload`, `params` and `body`, but `data` seemed best.

**Comment by timcappalli** [comment]
2024-09-18 call: good to merge once we have links

**Comment by marcoscaceres** [comment]
Updated the tests https://github.com/WICG/digital-credentials/pull/165

**Comment by marcoscaceres** [comment]
Filed issues on WebKit side...

**Comment by marcoscaceres** [comment]
Going ahead and merging, as we have tests and implementation commitments and it's mostly a cosmetic change.

**Comment by samuelgoto** [comment]
> Filed issues on WebKit side...

Chromium bug filed:

https://g-issues.chromium.org/issues/370982995

**Pull request merged**
